fdb-extensions/src/main/java/com/apple/foundationdb/annotation/GenerateVisitorAnnotationProcessor.java

```diff
@@ -0,0 +1,343 @@
+ * {@code accept()} method.
+ * </p><p>
+ * In cases where generic code takes care of the bulk of the logic but there may be some specialization for a few
+ * subclasses, a second interface called {@code ...WithDefaults} is created with default implementations of all
+ * visitation methods which just call {@code visitDefault()}. You can implement the generic logic in
+ * {@code visitDefault()}, implement the specific overrides by overriding the specific visitation method.
+ * </p>
+ */
+@AutoService(Processor.class)
+public class GenerateVisitorAnnotationProcessor extends AbstractProcessor {
+    private static final String parameterName = "element";
+
+    @Override
+    public synchronized void init(@Nonnull final ProcessingEnvironment processingEnv) {
+        super.init(processingEnv);
+    }
+
+    @Nonnull
+    @Override
+    public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.latestSupported();
+    }
+
+    @Override
+    public Set<String> getSupportedAnnotationTypes() {
+        return ImmutableSet.<String>builder()
+                .add(GenerateVisitor.class.getCanonicalName())
+                .build();
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        final var elementUtils = processingEnv.getElementUtils();
+        final var typeUtils = processingEnv.getTypeUtils();
+        final var messager = processingEnv.getMessager();
+        final var filer = processingEnv.getFiler();
+
+
+        for (final Element annotatedElement : roundEnv.getElementsAnnotatedWith(GenerateVisitor.class)) {
+            if (annotatedElement.getKind().isClass() && !annotatedElement.getKind().isInterface()) {
+                error(messager, annotatedElement, "only classes and interfaces can be annotated with %s", GenerateVisitor.class.getSimpleName());
+                return true;
+            }
+
+            final var rootTypeElement = (TypeElement) annotatedElement;
+
+            final var moduleElement = elementUtils.getModuleOf(annotatedElement);
+            if (moduleElement == null) {
+                error(messager, annotatedElement, "cannot annotate class with  %s in null-module", GenerateVisitor.class.getSimpleName());
+                return true;
+            }
+
+            if (!isValidClass(rootTypeElement)) {
+                error(messager,
+                        rootTypeElement, "The class %s cannot be annotated with this annotation.",
+                        rootTypeElement.getQualifiedName().toString());
+                return true;
+            }
+
+            final var generateVisitor = annotatedElement.getAnnotation(GenerateVisitor.class);
+            final var rootTypeMirror = rootTypeElement.asType();
+
+            final var packageOfRoot = elementUtils.getPackageOf(rootTypeElement);
+            final var subClassTypeMirrors = moduleElement
```

Comment:
How will it ever be possible to implement subclasses in other modules?

Reply:
Yes, you can do it by overriding `visitDefault()` and putting a switch in there.